### ISSUE TITLE
CRLF injection vulnerability

### ISSUE THREAD

**Issue by hannob** [urllib3/urllib3#1553]
At https://bugs.python.org/issue36276 there's an issue in Python's urllib that an attacker controlling the request parameter can inject headers by injecting CR/LF chars.

A commenter mentions that the same bug is present in urllib3:
https://bugs.python.org/issue36276#msg337837

So reporting it here to make sure it gets attention.

**Comment by haikuginger** [comment]
I confirmed that I was not able to reproduce as of the current master. Will investigate to see what the bisection point was.

**Comment by haikuginger** [comment]
Verified as fixed in #1487, but that change has not been released. I recommend we release 1.24.2 with that change as soon as possible, @sethmlarson @theacodes.

**Comment by theacodes** [comment]
I will try to cut a release soon, unless Seth wants to beat me to it. Thank
you, Jess.

**Comment by sethmlarson** [comment]
Without #1531 or IRI support in rfc3986 releasing master in it's current state will break backwards compatibility with international URLs.

**Comment by haikuginger** [comment]
I've done some checking and `requests` appears to have handling so indirect consumers via requests should not be vulnerable through us, if that's a useful datum to have.

**Comment by vstinner** [comment]
CVE-2016-5699, CVE-2019-9740 and CVE-2019-9947 have been assigned to HTTP Header Injection vulnerabilities in Python: see https://bugs.python.org/issue30458 for more information.

**Comment by Plazmaz** [comment]
I was able to reproduce this with `requests.get` using urllib, haven't tested with urllib3 though.

**Comment by carnil** [comment]
CVE-2019-11236 was assigned for this issue.

**Comment by rschultheis** [comment]
👋  Hi!  I've noticed that CVE-2019-11236 has been assigned to the CRLF injection issue described here.  It seems that the library has been patched in GitHub, but no new release has been made to pypi.  Will a new release containing the fix be made to pypi soon?    Based on @theacodes  [comment](https://github.com/urllib3/urllib3/issues/1553#issuecomment-474032699) it seems like a release was going to be made, but I also see her status has her perhaps unavailable.  Is someone else perhaps able to cut a new release into pypi?

**Comment by sethmlarson** [comment]
@rschultheis [Per my comment](https://github.com/urllib3/urllib3/issues/1553#issuecomment-474046652) we cannot cut a release until IRI support is fixed in rfc3986.

**Comment by rschultheis** [comment]
Thank you @sethmlarson ,  I will watch that PR 🙇

**Comment by vstinner** [comment]
"It seems that the library has been patched in GitHub"

Do you have a reference to the change?

**Comment by sethmlarson** [comment]
@vstinner 0aa3e24fcd75f1bb59ab159e9f8adb44055b2271

**Comment by luigino** [comment]
Is the problem fixed in 1.24.2? I couldn't find any reference to it in the changelog.

**Comment by sethmlarson** [comment]
No it is not, this change will be fixed in 1.25 and is next on my list of work.

**Comment by sethmlarson** [comment]
Closed by #1531. Will be released in 1.25

**Comment by r1b** [comment]
I contacted NVD to recommend that they update the CVE to reflect the fact that versions up to and including 1.24.2 are vulnerable - currently they only indicate that 1.24.1 is vulnerable.

See https://nvd.nist.gov/vuln/detail/CVE-2019-11236

**Comment by ryanpetrello** [comment]
Hello!

Is there any chance of this landing in a 1.24.3?  It would be nice to have a resolution to this CVE without necessarily having to inherit this change in 1.25.x:

https://github.com/urllib3/urllib3/pull/1507

**Comment by sethmlarson** [comment]
If that's the only change you're trying to circumvent and are willing to accept not verifying certificates you can still pass `cert_reqs="NONE"` into your `PoolManager` / `ConnectionPool`.

**Comment by ryanpetrello** [comment]
@sethmlarson that works so long as _your code_ is creating the manager/pool and you have direct influence over it (and your library isn't using some third-party dependency that imports `urllib3`.

**Comment by ryanpetrello** [comment]
@haikuginger I've also done some digging, and I don't see a way that requests' interfaces themselves are vulnerable to this exploit:

It looks like this specific code in `requests` foils the exploit by quoting the full URI:

```python
(Pdb) bt
  <stdin>(1)<module>()
  /home/ryanpetrello/venvs/tmp-56c32be26cf4d208/requests/requests/api.py(75)get()
  /home/ryanpetrello/venvs/tmp-56c32be26cf4d208/requests/requests/api.py(60)request()
  /home/ryanpetrello/venvs/tmp-56c32be26cf4d208/requests/requests/sessions.py(519)request()
  /home/ryanpetrello/venvs/tmp-56c32be26cf4d208/requests/requests/sessions.py(462)prepare_request()
  /home/ryanpetrello/venvs/tmp-56c32be26cf4d208/requests/requests/models.py(313)prepare()
> /home/ryanpetrello/venvs/tmp-56c32be26cf4d208/requests/requests/models.py(438)prepare_url()
(Pdb) l 435
430
431  	        enc_params = self._encode_params(params)
432  	        if enc_params:
433  	            if query:
434  	                query = '%s&%s' % (query, enc_params)
435  	            else:
436  	                query = enc_params
437
438  ->	        url = requote_uri(urlunparse([scheme, netloc, path, None, query, fragment]))
439  	        self.url = url
440
(Pdb) scheme, netloc, path, query, fragment
('http',
 'localhost:8000',
 '/ HTTP/1.1\r\nHEADER: INJECTED\r\nIgnore:',
 None,
 None)
```

Specifically:

```python
(Pdb) bt
  <stdin>(1)<module>()
  /home/ryanpetrello/venvs/tmp-56c32be26cf4d208/requests/requests/api.py(75)get()
  /home/ryanpetrello/venvs/tmp-56c32be26cf4d208/requests/requests/api.py(60)request()
  /home/ryanpetrello/venvs/tmp-56c32be26cf4d208/requests/requests/sessions.py(519)request()
  /home/ryanpetrello/venvs/tmp-56c32be26cf4d208/requests/requests/sessions.py(462)prepare_request()
  /home/ryanpetrello/venvs/tmp-56c32be26cf4d208/requests/requests/models.py(313)prepare()
  /home/ryanpetrello/venvs/tmp-56c32be26cf4d208/requests/requests/models.py(438)prepare_url()
> /home/ryanpetrello/venvs/tmp-56c32be26cf4d208/requests/requests/utils.py(603)requote_uri()
(Pdb) l 600
595  	    :rtype: str
596  	    """
597  	    safe_with_percent = "!#$%&'()*+,/:;=?@[]~"
598  	    safe_without_percent = "!#$&'()*+,/:;=?@[]~"
599  	    try:
600  	        # Unquote only the unreserved characters
601  	        # Then quote only illegal characters (do not quote reserved,
602  	        # unreserved, or '%')
603  ->	        return quote(unquote_unreserved(uri), safe=safe_with_percent)
604  	    except InvalidURL:
605  	        # We couldn't unquote the given URI, so let's try quoting it, but
(Pdb) uri
'http://localhost:8000/ HTTP/1.1\r\nHEADER: INJECTED\r\nIgnore:'
(Pdb) quote(unquote_unreserved(uri), safe=safe_with_percent)
'http://localhost:8000/%20HTTP/1.1%0D%0AHEADER:%20INJECTED%0D%0AIgnore:'
```

https://github.com/kennethreitz/requests/blob/v2.21.0/requests/utils.py#L603

**Comment by sethmlarson** [comment]
If we released a change that only checked the authority, path, and query for unquoted control characters would that be sufficient? The problem is is that pre-1.25 the `parse_url` function is really dumb, basically just splits on delimiters and goes with whatever.

**Comment by ryanpetrello** [comment]
@sethmlarson if something like that prevented the exploit described in `CVE-2019-11236`, it would be very much appreciated.

**Comment by sethmlarson** [comment]
I'm pretty sure that's what h11 did for their fix. I'll look at what they did and put a patch together.

**Comment by ryanpetrello** [comment]
Thanks @sethmlarson!

**Comment by ryanpetrello** [comment]
@sethmlarson do you have a link to said h11 fix?  I'd be glad to take a look and adapt a PR for `urllib3`.

**Comment by sethmlarson** [comment]
@ryanpetrello python-hyper/h11#69. I'll have to create a branch for `1.24-series` as the fix doesn't make sense to apply to `master`.

**Comment by ryanpetrello** [comment]
@sethmlarson I'm not sure which branch I'd want to submit a PR to, but how do you feel about this in general?

https://github.com/ryanpetrello/urllib3/commit/baf067b652a0e59a14c401edf5e59bfe39531424

**Comment by sethmlarson** [comment]
I would raise a `LocationParseError` instead of raising a `ValueError` but besides that it looks fine. Maybe we'd want to put a note that this is taken from cpython as well.  We'd also need some test cases for each control character.

**Comment by ryanpetrello** [comment]
@sethmlarson if you'll create 1.24 series branch, I'll open a PR based on this (with some tests).

**Comment by sethmlarson** [comment]
[I created the branch](https://github.com/urllib3/urllib3/tree/1.24-series).

**Comment by ryanpetrello** [comment]
An FYI for those in this thread: a resolution has also been merged into the 1.24 series and released as 1.24.3:

https://github.com/urllib3/urllib3/pull/1593
https://github.com/urllib3/urllib3/releases/tag/1.24.3

**Comment by MartinFalatic** [comment]
1.24.3 isn't on Pypi yet (hopefully soon?)
